src/display_context.cpp

```diff
@@ -87,21 +87,59 @@ display_context::can_move_result display_context::unit_can_move(const unit& u) c
 	const team& current_team = get_team(u.side());
 
 	can_move_result result = {false, false};
-	for(const map_location& adj : get_adjacent_tiles(u.get_location())) {
-		if (map().on_board(adj)) {
-			if(!result.attack_here) {
-				const unit_map::const_iterator i = units().find(adj);
-				if (i.valid() && !i->incapacitated() && current_team.is_enemy(i->side())) {
-					result.attack_here = true;
+
+	// Generating pairs of min-max ranges of all units weapons
+	const auto& attacks = u.attacks();
+	std::vector<std::pair<int, int>> distance_ranges;
+	for (const auto& attack : attacks) {
+		int min_range = attack.min_range();
+		int max_range = attack.max_range();
+		distance_ranges.emplace_back(min_range , max_range);
+	}
+
+	// Merging it here, may be removed if some checks for every weapon
+	// Should be done individually (Like line-of sight)
+	std::sort(distance_ranges.begin(), distance_ranges.end());
+
+	std::vector<std::pair<int, int>> merged_ranges;
+	for (const auto& range : distance_ranges) {
+		if (merged_ranges.empty() || merged_ranges.back().second < range.first) {
+			merged_ranges.push_back(range);
+		} else {
+			merged_ranges.back().second = std::max(merged_ranges.back().second, range.second);
+		}
+	}
+
+	for (const auto& range : merged_ranges) {
+		int min_distance = range.first;
+		int max_distance = range.second;
```

Comment:
I think you can avoid this outermost loop by putting every valid range into a set, and for each hex check that the distance is in that set.

For example, if you have weapons with the following ranges:

* min_range=2, max_range=3
* min_range=6, max_range=8
* min_range=2, max_range=4
* min_range=3, max_range=4

Then your set of valid ranges contains the following values:

`{2, 3, 4, 6, 7, 8}`

Then, assuming your set is non-empty, this becomes:

```cpp
	int min_distance = *valid_ranges.begin();
	int max_distance = *std::prev(valid_ranges.end());
```

Reply:
True, I think.